mvp-spec/website-design.md

```diff
@@ -0,0 +1,130 @@
+  - depending on the context, this operator client will call a JSON or "redirect" endpoint on the operator
+
+### Test support of 3rd party cookies
+
+The logic to test if third party cookies are supported and use full page redirect if they aren't is the following:
+
+1. if an HTTP middleware is used, then
+   1. based on user agent, if the browser is known to **not** support 3PC (ex: Safari) ➡️ consider no 3PC and immediately **HTTP redirect**
+2. in Javascript,
+   1. based on user agent, if the browser is known to **not** support 3PC (ex: Safari) ➡️ consider no 3PC and immediately **javascript redirect**
+3. otherwise, attempt to call JSON endpoint and read existing (3PC) Prebid ID
+   1. at the same time, the operator attempts to write a "test" cookie on .prebidsso.com (ie. attempt to write a 3PC)
+4. if Prebid ID is retrieved, of course it means 3PC **are** supported
+5. if no Prebid ID retrieved, it means either that the user is not known, or that 3PC are not supported
+   1. call operator to attempt to read "test" cookie that was just written
+   2. if success ➡️ 3PC **are** supported, it's just that the user is not known
+   3. if failure ➡️ 3PC are **not** supported, **javascript redirect**
+
+```mermaid
+flowchart TD
+    subgraph noMiddleware [HTTP server without middleware]
+        NoMiddlewareServe[serve HTML page]
+    end
+    
+    subgraph middleware[HTTP server with middleware]
+        MiddlewareCookies{"Any Prebid 1st party 🍪?"}
+        MiddlewareAfterRedirect{Redirected from operator?}
+        MiddlewareNonEmptyData{Received data?}
+        Middleware3PC{Browser known to support 3PC?}
+        MiddlewareServe[serve HTML page]
+        MiddlewareRedirect[HTTP redirect]
+        MiddlewareSave["Set 1st party Prebid 🍪 = data"]
+        MiddlewareSaveNothing["Set 1st party Prebid 🍪 = 'unknown'"]
+        
+        Middleware3PC -->|No| MiddlewareRedirect
+        Middleware3PC -->|Yes| MiddlewareServe
+        
+        MiddlewareCookies -->|No| MiddlewareAfterRedirect
+        
+        MiddlewareAfterRedirect -->|No| Middleware3PC
+        MiddlewareAfterRedirect -->|Yes| MiddlewareNonEmptyData
+        
+        MiddlewareNonEmptyData -->|No| MiddlewareSaveNothing
+        MiddlewareNonEmptyData -->|Yes| MiddlewareSave
+        
+        MiddlewareSaveNothing --> MiddlewareServe
+        MiddlewareSave --> MiddlewareServe
+        
+        MiddlewareCookies -->|Yes| MiddlewareServe
+    end
+
+    subgraph "HTML page (Javascript)"
+        HtmlLoad[Page load]
+        HtmlAfterRedirect{Redirected from operator?}
+        HtmlCookies{"Any Prebid 1st party 🍪?"}
+        HtmlSave["Set 1st party Prebid 🍪 = data"]
+        HtmlSaveNothing["Set 1st party Prebid 🍪 = 'unknown'"]
+        HtmlAnyData{Received data?}
+        HtmlDone[Finished]
+        
+        ClientBrowser3PC{Browser known to support 3PC?}
+        ClientNonEmptyData{Received data?}
+        Client3PCOk{3d party cookie ok?}
+        
+    end
```

Comment:
Part of the flow is not clear for me. Particularly with the "Redirected from the operator?" that is both in the "HTTP server with middleware" and the "HTML page (Javascript)" considering that it is possible to start the flow in the former and then move to the latest. If so, we potentially "Set 1st party Prebid 🍪 = data" two times.

I am not sure, but I think that the flow can be resplit in the chart to avoid this ambiguity.

Reply:
Actually, if the redirection is handled by the HTTP middleware, then cookies are set, and in the HTML (javascript) section, you wouldn't end up in this same branch because to the question "any 1st party data", the answer is "yes".
Do you see a way to clarify this?